File: pipeline/rpc.py
"""Batch JSON-RPC client for the Pons factory. stdlib only (urllib). This is
the only module that opens a network connection anywhere in the pipeline.

RPC envelope (ARCHITECTURE.md section 1): batch size 50, 2.0s pacing between
batches for eth_getBlockByNumber/eth_call. A 429 arrives as a single object
where an array was requested -- treated as one retryable failure, not a
parse error. eth_getLogs keeps crawl0.py's proven single-request pacing,
paced by the caller (pipeline/crawl.py), not here.
"""
from __future__ import annotations

import json
import socket
import time
import urllib.error
import urllib.request
from typing import Callable, Optional

TOPIC_TOKEN_LAUNCHED = "0x8d4aad4953d0ca700d468f3753aa14432d1b35b43ec6409f051fb6aa43a89607"
TOPIC_POOL_GRADUATED = "0x0a44ef75df69c534f43cd6c1aa3ef8983065fe5fe79ef9e79f6494e6f258c259"

FACTORY_ADDRESS = "0x7eD598BcEf8bd9Edd8C97A195C6d13f40801EC7e"
SYMBOL_SELECTOR = "0x95d89b41"  # symbol()

MAX_BATCH = 50
BATCH_PACING_SECONDS = 2.0
MAX_RETRIES = 9
BACKOFF_BASE_SECONDS = 2.0
BACKOFF_CAP_SECONDS = 120.0
RETRYABLE_HTTP = (408, 429, 500, 502, 503, 504)

Transport = Callable[[list], object]


def _pad32(hexstr: str) -> str:
    return hexstr.rjust(64, "0")


def _topic_to_address(topic: str) -> str:
    return "0x" + topic[-40:]


def _data_word(data: str, index: int) -> str:
    start = 2 + index * 64
    return data[start : start + 64]


def is_retryable(response: object) -> bool:
    """Transport-level failures (HTTP 429/5xx, timeouts) the transport hands
    back as a single object; retried with the same backoff as a 429."""
    return isinstance(response, dict) and response.get("code") in RETRYABLE_HTTP


def is_rate_limited(response: object) -> bool:
    """A 429 comes back as a single JSON object instead of the requested
    array. Any other single-object response is a real error, not a
    rate limit. Narrows is_retryable to the one code that means "slow down".
    """
    return is_retryable(response) and response.get("code") == 429


class MalformedBatchResponse(Exception):
    """The provider answered a batch with something that cannot be matched
    back to the requests (wrong length, missing/duplicate ids). Retried on
    the same backoff path as a 429: it is a transport fault, never data."""


def index_batch_response(response: object, expected: int) -> list:
    """Match a JSON-RPC batch response back to its requests by `id`.

    JSON-RPC 2.0 lets a server return batch responses in any order, and a
    short array would otherwise be zipped positionally against the requests
    -- shifting every later result onto the wrong block. Both are refused.
    """
    if not isinstance(response, list):
        raise MalformedBatchResponse(f"expected an array of {expected} responses, got {type(response).__name__}")
    if len(response) != expected:
        raise MalformedBatchResponse(f"expected {expected} responses, got {len(response)}")
    by_id: dict = {}
    for item in response:
        if not isinstance(item, dict) or "id" not in item:
            raise MalformedBatchResponse("batch response item carries no id")
        if item["id"] in by_id:
            raise MalformedBatchResponse(f"duplicate id {item['id']!r} in batch response")
        by_id[item["id"]] = item
    missing = [i for i in range(expected) if i not in by_id]
    if missing:
        raise MalformedBatchResponse(f"batch response missing ids {missing}")
    return [by_id[i].get("result") for i in range(expected)]


def decode_token_launched(log: dict) -> dict:
    topics = log["topics"]
    data = log["data"]
    return {
        "token": _topic_to_address(topics[1]),
        "curve": _topic_to_address(topics[2]),
        "deployer": _topic_to_address(topics[3]),
        "pairToken": "0x" + _data_word(data, 0)[-40:],
        "launchConfigId": int(_data_word(data, 1), 16),
        "graduationThreshold": int(_data_word(data, 2), 16),
        "block": int(log["blockNumber"], 16),
        "txHash": log["transactionHash"],
        "logIndex": int(log["logIndex"], 16),
    }


def decode_pool_graduated(log: dict) -> dict:
    topics = log["topics"]
    data = log["data"]
    return {
        "token": _topic_to_address(topics[1]),
        "positionId": int(_data_word(data, 0), 16),
        "tokenAmount": int(_data_word(data, 1), 16),
        "pairTokenAmount": int(_data_word(data, 2), 16),
        "block": int(log["blockNumber"], 16),
        "txHash": log["transactionHash"],
        "logIndex": int(log["logIndex"], 16),
    }


def _decode_string_return(hex_data: str) -> Optional[str]:
    raw = bytes.fromhex(hex_data[2:])
    if len(raw) < 64:
        return None
    length = int.from_bytes(raw[32:64], "big")
    return raw[64 : 64 + length].decode("utf-8", errors="replace")


def _urllib_transport(url: str) -> Transport:
    def send(payload: list) -> object:
        body = json.dumps(payload).encode()
        request = urllib.request.Request(url, data=body, headers={"Content-Type": "application/json", "User-Agent": "ledge/1.0 (+https://ledge.tools)"})
        try:
            with urllib.request.urlopen(request, timeout=45) as response:
                return json.loads(response.read())
        except urllib.error.HTTPError as exc:
            if exc.code in RETRYABLE_HTTP:
                return {"code": exc.code, "message": str(exc)}
            raise
        except (urllib.error.URLError, socket.timeout, TimeoutError) as exc:
            return {"code": 503, "message": f"transport: {exc}"}

    return send


class RpcClient:
    """Batch JSON-RPC client with 429-as-object detection and backoff.
    `transport` is injectable for tests; production uses urllib."""

    def __init__(self, url: str, transport: Optional[Transport] = None):
        self.url = url
        self.transport = transport or _urllib_transport(url)

    def _send_with_retry(self, payload: list, validate: Optional[Callable[[object], list]] = None) -> list:
        """Send one payload, retrying transport faults with exponential
        backoff. `validate` maps a raw response to the value to return and
        may raise MalformedBatchResponse to send the request down that same
        retry path."""
        delay = BACKOFF_BASE_SECONDS
        for attempt in range(MAX_RETRIES):
            response = self.transport(payload)
            if is_retryable(response):
                if attempt == MAX_RETRIES - 1:
                    raise RuntimeError(f"rpc: gave up after repeated failures: {response}")
                time.sleep(delay)
                delay = min(delay * 2, BACKOFF_CAP_SECONDS)
                continue
            if validate is None:
                return response
            try:
                return validate(response)
            except MalformedBatchResponse as exc:
                if attempt == MAX_RETRIES - 1:
                    raise RuntimeError(f"rpc: malformed batch response after retries: {exc}") from exc
                time.sleep(delay)
                delay = min(delay * 2, BACKOFF_CAP_SECONDS)
        raise RuntimeError("rpc: gave up after repeated 429 responses")

    def call_batch(self, requests: list) -> list:
        """Results are matched back to requests by `id`, never positionally.
        A response that is short, over-long, or missing an id is refused and
        retried; it never reaches the caller as a shifted or truncated list.
        """
        results: list = []
        chunks = [requests[i : i + MAX_BATCH] for i in range(0, len(requests), MAX_BATCH)]
        for chunk_index, chunk in enumerate(chunks):
            payload = [
                {"jsonrpc": "2.0", "id": i, "method": r["method"], "params": r["params"]}
                for i, r in enumerate(chunk)
            ]
            expected = len(chunk)
            results.extend(
                self._send_with_retry(payload, validate=lambda r, n=expected: index_batch_response(r, n))
            )
            if chunk_index < len(chunks) - 1:
                time.sleep(BATCH_PACING_SECONDS)
        return results

    def get_head_block(self) -> int:
        payload = [{"jsonrpc": "2.0", "id": 0, "method": "eth_blockNumber", "params": []}]
        response = self._send_with_retry(payload)
        return int(response[0]["result"], 16)

    def get_logs(self, from_block: int, to_block: int, topic0: str) -> list:
        payload = [
            {
                "jsonrpc": "2.0",
                "id": 0,
                "method": "eth_getLogs",
                "params": [
                    {
                        "fromBlock": hex(from_block),
                        "toBlock": hex(to_block),
                        "address": FACTORY_ADDRESS,
                        "topics": [topic0],
                    }
                ],
            }
        ]
        # A per-item error (e.g. "log query timed out") -- or an item with no
        # "result" at all -- must never be read as an empty window: only an
        # explicit result is an answer. Anything else retries, then fails
        # the run.
        delay = BACKOFF_BASE_SECONDS
        for attempt in range(MAX_RETRIES):
            response = self._send_with_retry(payload)
            item = response[0] if isinstance(response, list) and response else None
            if isinstance(item, dict) and "error" not in item and "result" in item:
                return item["result"] or []
            if attempt == MAX_RETRIES - 1:
                raise RuntimeError(f"rpc: eth_getLogs gave no result after retries: {item!r}")
            time.sleep(delay)
            delay *= 2
        raise RuntimeError("rpc: unreachable")

    def symbol_of(self, address: str) -> Optional[str]:
        payload = [
            {
                "jsonrpc": "2.0",
                "id": 0,
                "method": "eth_call",
                "params": [{"to": address, "data": SYMBOL_SELECTOR}, "latest"],
            }
        ]
        response = self._send_with_retry(payload)
        result = response[0].get("result")
        return _decode_string_return(result) if result else None

File: worker/src/rpc.ts
/* Batch JSON-RPC, a port of pipeline/rpc.py's proven envelope.

   The four things that are not negotiable, each learned the hard way and each
   restated here because the Worker shares an RPC with the hourly Python:

     1. `User-Agent: ledge/1.0` is REQUIRED. Without it the endpoint answers
        403 to every request.
     2. Batches are at most 50 requests, with 2.0 s between batches.
     3. A 429 arrives as a single object where an array was requested. That is
        a transport fault, never data, and it is retried -- or, in the tick,
        skipped so the next minute covers a wider window.
     4. Batch responses are matched back to requests by `id`, never
        positionally. A short or reordered array would shift every later
        result onto the wrong block.

   eth_getLogs windows stay at 1,000 blocks: the public endpoint rate-limits
   anything wider (PONS_CONTRACTS.md, crawl proof). */

export const MAX_BATCH = 50;
export const BATCH_PACING_MS = 2000;
export const LOG_WINDOW_BLOCKS = 1000;
export const USER_AGENT = "ledge/1.0 (+https://ledge.tools)";
const RETRYABLE_HTTP = new Set([408, 429, 500, 502, 503, 504]);
const MAX_RETRIES = 4;
const BACKOFF_BASE_MS = 1000;
const BACKOFF_CAP_MS = 8000;
const TIMEOUT_MS = 20_000;

export class RpcUnavailable extends Error {
  readonly rateLimited: boolean;
  constructor(message: string, rateLimited = false) {
    super(message);
    this.name = "RpcUnavailable";
    this.rateLimited = rateLimited;
  }
}

export class MalformedBatchResponse extends Error {}

export interface RpcRequest {
  method: string;
  params: unknown[];
}

interface TransportFault {
  code: number;
  message: string;
}

/** Injectable so tests never touch the network. */
export type Transport = (payload: unknown[]) => Promise<unknown>;

function isFault(response: unknown): response is TransportFault {
  return (
    typeof response === "object" &&
    response !== null &&
    !Array.isArray(response) &&
    typeof (response as TransportFault).code === "number" &&
    RETRYABLE_HTTP.has((response as TransportFault).code)
  );
}

export function isRateLimited(response: unknown): boolean {
  return isFault(response) && response.code === 429;
}

/** Match a batch response back to its requests by `id`. Anything that cannot
    be matched is refused rather than zipped positionally. */
export function indexBatchResponse(response: unknown, expected: number): unknown[] {
  if (!Array.isArray(response)) {
    throw new MalformedBatchResponse(
      `expected an array of ${expected} responses, got ${typeof response}`,
    );
  }
  if (response.length !== expected) {
    throw new MalformedBatchResponse(`expected ${expected} responses, got ${response.length}`);
  }
  const byId = new Map<number, Record<string, unknown>>();
  for (const item of response) {
    if (typeof item !== "object" || item === null || !("id" in item)) {
      throw new MalformedBatchResponse("batch response item carries no id");
    }
    const id = (item as { id: number }).id;
    if (byId.has(id)) throw new MalformedBatchResponse(`duplicate id ${id} in batch response`);
    byId.set(id, item as Record<string, unknown>);
  }
  const out: unknown[] = [];
  for (let i = 0; i < expected; i++) {
    const item = byId.get(i);
    if (item === undefined) throw new MalformedBatchResponse(`batch response missing id ${i}`);
    out.push(item["result"]);
  }
  return out;
}

function httpTransport(url: string): Transport {
  return async (payload) => {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
    try {
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          // Required. The endpoint answers 403 without it.
          "User-Agent": USER_AGENT,
        },
        body: JSON.stringify(payload),
        signal: controller.signal, // lint-copy:allow — AbortController, not copy
      });
      if (!response.ok) {
        if (RETRYABLE_HTTP.has(response.status)) {
          return { code: response.status, message: `HTTP ${response.status}` };
        }
        throw new RpcUnavailable(`rpc: HTTP ${response.status}`);
      }
      return await response.json();
    } catch (error) {
      if (error instanceof RpcUnavailable) throw error;
      return { code: 503, message: `transport: ${String(error)}` };
    } finally {
      clearTimeout(timer);
    }
  };
}

const sleep = (ms: number) => new Promise<void>((resolve) => setTimeout(resolve, ms));

export class RpcClient {
  private readonly transport: Transport;
  /** True once any request in this client's lifetime hit a 429. The tick reads
      it to decide whether to hold the cursor rather than record a failure. */
  rateLimitSeen = false;

  /** A second endpoint tried only after the first has given up. The
      official RPC rate-limits Cloudflare's shared egress addresses hard;
      a fallback keeps a lookup honest-but-answered instead of rpc_down. */
  private readonly fallback: Transport | null;

  constructor(url: string, transport?: Transport, fallbackUrl?: string, fallbackTransport?: Transport) {
    this.transport = transport ?? httpTransport(url);
    this.fallback = fallbackTransport ?? (fallbackUrl && fallbackUrl !== url ? httpTransport(fallbackUrl) : null);
  }

  private async sendWithRetry(
    payload: unknown[],
    validate?: (response: unknown) => unknown[],
  ): Promise<unknown> {
    try {
      return await this.sendVia(this.transport, payload, validate);
    } catch (error) {
      if (!(error instanceof RpcUnavailable) || this.fallback === null) throw error;
      return await this.sendVia(this.fallback, payload, validate);
    }
  }

  private async sendVia(
    transport: Transport,
    payload: unknown[],
    validate?: (response: unknown) => unknown[],
  ): Promise<unknown> {
    let delay = BACKOFF_BASE_MS;
    for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
      const response = await transport(payload);
      if (isFault(response)) {
        if (response.code === 429) this.rateLimitSeen = true;
        if (attempt === MAX_RETRIES - 1) {
          throw new RpcUnavailable(`rpc: gave up: ${response.message}`, response.code === 429);
        }
        await sleep(delay);
        delay = Math.min(delay * 2, BACKOFF_CAP_MS);
        continue;
      }
      if (!validate) return response;
      try {
        return validate(response);
      } catch (error) {
        if (!(error instanceof MalformedBatchResponse)) throw error;
        if (attempt === MAX_RETRIES - 1) {
          throw new RpcUnavailable(`rpc: malformed batch response: ${error.message}`);
        }
        await sleep(delay);
        delay = Math.min(delay * 2, BACKOFF_CAP_MS);
      }
    }
    throw new RpcUnavailable("rpc: gave up after repeated transport faults");
  }

  /** Chunked at MAX_BATCH with BATCH_PACING_MS between chunks. */
  async callBatch(requests: RpcRequest[]): Promise<unknown[]> {
    const results: unknown[] = [];
    const chunks: RpcRequest[][] = [];
    for (let i = 0; i < requests.length; i += MAX_BATCH) {
      chunks.push(requests.slice(i, i + MAX_BATCH));
    }
    for (let c = 0; c < chunks.length; c++) {
      const chunk = chunks[c] as RpcRequest[];
      const payload = chunk.map((r, i) => ({
        jsonrpc: "2.0",
        id: i,
        method: r.method,
        params: r.params,
      }));
      const expected = chunk.length;
      const part = (await this.sendWithRetry(payload, (response) =>
        indexBatchResponse(response, expected),
      )) as unknown[];
      results.push(...part);
      if (c < chunks.length - 1) await sleep(BATCH_PACING_MS);
    }
    return results;
  }

  async getHeadBlock(): Promise<number> {
    const response = (await this.sendWithRetry([
      { jsonrpc: "2.0", id: 0, method: "eth_blockNumber", params: [] },
    ])) as Array<{ result?: string }>;
    const head = response[0]?.result;
    if (!head) throw new RpcUnavailable("rpc: eth_blockNumber gave no result");
    return Number(BigInt(head));
  }

  /** A per-item error, or an item with no `result` at all, must never be read
      as an empty window: only an explicit result is an answer. */
  /** `address` is nullable on purpose. Factory events (TokenLaunched,
      PoolGraduated) are filtered by the factory address and must be. Curve
      events cannot be: every launch has its own curve, so there is no address
      to filter by and the filter is topic0 alone, with the emitting curve read
      out of each log's own `address` field (RESEARCH-PHASE2-3.md section 0). */
  async getLogs(
    fromBlock: number,
    toBlock: number,
    address: string | null,
    topic0: string,
  ): Promise<unknown[]> {
    const payload = [
      {
        jsonrpc: "2.0",
        id: 0,
        method: "eth_getLogs",
        params: [
          {
            fromBlock: "0x" + fromBlock.toString(16),
            toBlock: "0x" + toBlock.toString(16),
            address,
            topics: [topic0],
          },
        ],
      },
    ];
    const response = (await this.sendWithRetry(payload)) as Array<Record<string, unknown>>;
    const item = Array.isArray(response) ? response[0] : undefined;
    if (item && !("error" in item) && "result" in item) {
      return (item["result"] as unknown[]) ?? [];
    }
    throw new RpcUnavailable(`rpc: eth_getLogs gave no result: ${JSON.stringify(item)}`);
  }

  async ethCall(to: string, data: string): Promise<string | null> {
    const results = await this.callBatch([
      { method: "eth_call", params: [{ to, data }, "latest"] },
    ]);
    const result = results[0];
    return typeof result === "string" ? result : null;
  }
}
